src/avx512-64bit-keyvaluesort.hpp

```diff
@@ -0,0 +1,1276 @@
+    vtype::mask_storeu(keys + 88, load_mask4, key_zmm[11]);
+    vtype::mask_storeu(keys + 96, load_mask5, key_zmm[12]);
+    vtype::mask_storeu(keys + 104, load_mask6, key_zmm[13]);
+    vtype::mask_storeu(keys + 112, load_mask7, key_zmm[14]);
+    vtype::mask_storeu(keys + 120, load_mask8, key_zmm[15]);
+}
+
+template <typename vtype, typename type_t>
+X86_SIMD_SORT_INLINE type_t get_pivot_64bit(type_t *keys,
+                                            uint64_t *indexes,
+                                            const int64_t left,
+                                            const int64_t right)
+{
+    // median of 8
+    int64_t size = (right - left) / 8;
+    using zmm_t = typename vtype::zmm_t;
+    using index_t = zmm_kv_vector<uint64_t>::zmm_t;
+    __m512i rand_index = _mm512_set_epi64(left + size,
+                                          left + 2 * size,
+                                          left + 3 * size,
+                                          left + 4 * size,
+                                          left + 5 * size,
+                                          left + 6 * size,
+                                          left + 7 * size,
+                                          left + 8 * size);
+    zmm_t key_vec = vtype::template i64gather<sizeof(type_t)>(rand_index, keys);
+
+    index_t index_vec;
+    zmm_t sort;
+    index_vec = zmm_kv_vector<uint64_t>::template i64gather<sizeof(uint64_t)>(
+            rand_index, indexes);
+    sort = sort_zmm_64bit<vtype>(key_vec, index_vec);
+    // pivot will never be a nan, since there are no nan's!
+
+    return ((type_t *)&sort)[4];
+}
+
+template <typename vtype, typename type_t>
+inline void heapify(type_t *keys, uint64_t *indexes, int64_t idx, int64_t size)
+{
+    int64_t i = idx;
+    while (true) {
+        int64_t j = 2 * i + 1;
+        if (j >= size || j < 0) { break; }
+        int k = j + 1;
+        if (k < size && keys[j] < keys[k]) { j = k; }
+        if (keys[j] < keys[i]) { break; }
+        std::swap(keys[i], keys[j]);
+        std::swap(indexes[i], indexes[j]);
+        i = j;
+    }
+}
+template <typename vtype, typename type_t>
+inline void heap_sort(type_t *keys, uint64_t *indexes, int64_t size)
+{
+    for (int64_t i = size / 2 - 1; i >= 0; i--) {
+        heapify<vtype>(keys, indexes, i, size);
+    }
+    for (int64_t i = size - 1; i > 0; i--) {
+        std::swap(keys[0], keys[i]);
+        std::swap(indexes[0], indexes[i]);
+        heapify<vtype>(keys, indexes, 0, i);
+    }
+}
+
+template <typename vtype, typename type_t>
+inline void qsort_64bit_(type_t *keys,
+                         uint64_t *indexes,
+                         int64_t left,
+                         int64_t right,
+                         int64_t max_iters)
+{
+    /*
+     * Resort to std::sort if quicksort isnt making any progress
+     */
+    if (max_iters <= 0) {
+        heap_sort<vtype>(keys + left, indexes + left, right - left + 1);
```

Comment:
why not use std::sort? Just like how you used it in test_all.cpp.

Reply:
Because the arguments are two column pointer(keys and values), if we want to use std::sort, the input of std::sort is a struct, not two arguments, we need create a new two-dimension vector to store key and index, it need to malloc extra space.